### ISSUE TITLE
ipfs-update fails installing with ERROR: version didnt match 

### ISSUE THREAD

**Issue by mjrider** [ipfs/ipfs-update#50]
tried to install ipfs with the latest ipfs-update binary from dist.ipfs.io on a clean debian stable vm

Versions i tried:
- v0.4.1
- v0.4.2
- v0.4.3
- v0.4.5-pre1
- v0.4.4

all result in the same error:
```
daemon check failed: open /home/vagrant/.ipfs/api: no such file or directory
no pre-existing ipfs installation found
fetching go-ipfs version v0.4.4
  - using GOOS=linux and GOARCH=amd64
  - fetching "/ipns/dist.ipfs.io/go-ipfs/v0.4.4/go-ipfs_v0.4.4_linux-amd64.tar.gz"
fetching url: https://ipfs.io/ipns/dist.ipfs.io/go-ipfs/v0.4.4/go-ipfs_v0.4.4_linux-amd64.tar.gz
  - writing to /tmp/ipfs-update592994973/go-ipfs_v0.4.4_linux-amd64.tar.gz
  - extracting binary to tempdir:  /tmp/ipfs-update645384374/ipfs-new
binary downloaded, verifying...
  - running init in '/home/vagrant/.ipfs/update-staging/test306772824' with new binary
  - running: [/tmp/ipfs-update645384374/ipfs-new init]
  - checking new binary outputs correct version
  - running: [/tmp/ipfs-update645384374/ipfs-new version]
install failed, reverting changes...
ERROR: version didnt match 
```

if i run the last command by hand:
```
vagrant@jessie:~$ /tmp/ipfs-update645384374/ipfs-new version
ipfs version 0.4.4
```

**Comment by whyrusleeping** [comment]
which version of ipfs-update are you using? It looks like the daemon its trying to start up for verification is failing to come online.

Could you try setting `IPFS_PATH` to `/home/vagrant/.ipfs/update-staging/test306772824` and then running `/tmp/ipfs-update645384374/ipfs-new daemon`?

If that works fine, then i'll have to go dive into the issue a bit more.

Aside from debugging the binary validation process though, You should be able to complete the install by passing the `--no-check` flag to the install command.

**Comment by mjrider** [comment]
i used: ipfs-update_v1.5.0_linux-amd64.tar.gz from dist.ipfs.io

i already threw the vm away, so the paths do not match with  the output above

```
vagrant@test:~$ ipfs-update/ipfs-update --verbose install v0.4.4
daemon check failed: open /home/vagrant/.ipfs/api: no such file or directory
no pre-existing ipfs installation found
fetching go-ipfs version v0.4.4
  - using GOOS=linux and GOARCH=amd64
  - fetching "/ipns/dist.ipfs.io/go-ipfs/v0.4.4/go-ipfs_v0.4.4_linux-amd64.tar.gz"
fetching url: https://ipfs.io/ipns/dist.ipfs.io/go-ipfs/v0.4.4/go-ipfs_v0.4.4_linux-amd64.tar.gz
  - writing to /tmp/ipfs-update830245118/go-ipfs_v0.4.4_linux-amd64.tar.gz
  - extracting binary to tempdir:  /tmp/ipfs-update213901147/ipfs-new
binary downloaded, verifying...
  - running init in '/home/vagrant/.ipfs/update-staging/test746413637' with new binary
  - running: [/tmp/ipfs-update213901147/ipfs-new init]
  - checking new binary outputs correct version
  - running: [/tmp/ipfs-update213901147/ipfs-new version]
install failed, reverting changes...
ERROR: version didnt match 
vagrant@test:~$ export IPFS_PATH=/home/vagrant/.ipfs/update-staging/test746413637
vagrant@test:~$ /tmp/ipfs-update213901147/ipfs-new daemon
Initializing daemon...
Error: no ipfs repo found in /home/vagrant/.ipfs/update-staging/test746413637.
please run: ipfs init

```
ran init manual:
```
vagrant@test:~$ /tmp/ipfs-update213901147/ipfs-new init
initializing ipfs node at /home/vagrant/.ipfs/update-staging/test746413637
generating 2048-bit RSA keypair...done
peer identity: QmTB96ZN6dgbd3JWx7DYP8e8Ztz9ZijkyGe7Fsm7CqSpYz
to get started, enter:

	ipfs cat /ipfs/QmYwAPJzv5CZsnA625s3Xf2nemtYgPpHdWEz79ojWnPbdG/readme

vagrant@test:~$ /tmp/ipfs-update213901147/ipfs-new daemon
Initializing daemon...
Swarm listening on /ip4/10.0.3.96/tcp/4001
Swarm listening on /ip4/127.0.0.1/tcp/4001
Swarm listening on /ip6/::1/tcp/4001
```


afterwards, i removed the .ipfs directory, and tried again. but after the failure i checked the contents of the tempdir (/home/vagrant/.ipfs/update-staging/test746413637) and it was empty. it looks like the init isn't creating the expected files.

**Comment by mjrider** [comment]
the --no-check options is also not working

```
vagrant@test:~$ ipfs-update/ipfs-update --verbose install --no-check v0.4.4
daemon check failed: open /home/vagrant/.ipfs/api: no such file or directory
no pre-existing ipfs installation found
fetching go-ipfs version v0.4.4
  - using GOOS=linux and GOARCH=amd64
  - fetching "/ipns/dist.ipfs.io/go-ipfs/v0.4.4/go-ipfs_v0.4.4_linux-amd64.tar.gz"
fetching url: https://ipfs.io/ipns/dist.ipfs.io/go-ipfs/v0.4.4/go-ipfs_v0.4.4_linux-amd64.tar.gz
  - writing to /tmp/ipfs-update881005430/go-ipfs_v0.4.4_linux-amd64.tar.gz
  - extracting binary to tempdir:  /tmp/ipfs-update352090035/ipfs-new
skipping tests since '--no-check' was passed
skipping stash, no previous install
install failed, reverting changes...
ERROR: could not find good install location 
vagrant@test:~$ ipfs-update/ipfs-update --verbose install --no-check v0.4.2
daemon check failed: open /home/vagrant/.ipfs/api: no such file or directory
no pre-existing ipfs installation found
fetching go-ipfs version v0.4.2
  - using GOOS=linux and GOARCH=amd64
  - fetching "/ipns/dist.ipfs.io/go-ipfs/v0.4.2/go-ipfs_v0.4.2_linux-amd64.tar.gz"
fetching url: https://ipfs.io/ipns/dist.ipfs.io/go-ipfs/v0.4.2/go-ipfs_v0.4.2_linux-amd64.tar.gz
  - writing to /tmp/ipfs-update929058328/go-ipfs_v0.4.2_linux-amd64.tar.gz
  - extracting binary to tempdir:  /tmp/ipfs-update104671837/ipfs-new
skipping tests since '--no-check' was passed
skipping stash, no previous install
install failed, reverting changes...
ERROR: could not find good install location 
```

**Comment by skippa** [comment]
Same problem, following

**Comment by haniawni** [comment]
Yep, same problem on an Ubuntu 16.04 clean DO Droplet. Getting "could not find good install location" error.

```
syncthing@syncthing:~/gocode$ ipfs-update --verbose install latest
  - fetching "/ipns/dist.ipfs.io/go-ipfs/versions"
fetching url: https://ipfs.io/ipns/dist.ipfs.io/go-ipfs/versions
daemon check failed: open /home/syncthing/.ipfs/api: no such file or directory
no pre-existing ipfs installation found
fetching go-ipfs version v0.4.10
  - using GOOS=linux and GOARCH=amd64
  - fetching "/ipns/dist.ipfs.io/go-ipfs/v0.4.10/go-ipfs_v0.4.10_linux-amd64.tar.gz"
fetching url: https://ipfs.io/ipns/dist.ipfs.io/go-ipfs/v0.4.10/go-ipfs_v0.4.10_linux-amd64.tar.gz
  - writing to /tmp/ipfs-update822360374/go-ipfs_v0.4.10_linux-amd64.tar.gz
  - extracting binary to tempdir:  /tmp/ipfs-update364755571/ipfs-new
binary downloaded, verifying...
  - running init in '/home/syncthing/.ipfs/update-staging/test812963613' with new binary
  - running: [/tmp/ipfs-update364755571/ipfs-new init]
  - checking new binary outputs correct version
  - running: [/tmp/ipfs-update364755571/ipfs-new version]
  - tweaking test config to avoid external interference
  - starting up daemon
  - waiting on daemon to come online
  - found api file: 127.0.0.1:45145
  - Successfully made connection to api endpoint
  - checking that we can add and cat a file
  - running: [/tmp/ipfs-update364755571/ipfs-new cat QmTFJQ68kaArzsqz2Yjg1yMyEA5TXTfNw6d9wSFhxtBxz2]
  - checking that file shows up in ipfs refs local
success! tests all passed.
  - killing test daemon
skipping stash, no previous install
install failed, reverting changes...
ERROR: could not find good install location
```

**Comment by haniawni** [comment]
yea, if your bin dir of choice is in a different location than either $GOPATH/bin or $HOME/bin or /usr/bin, then it won't find it- mine is ~/.local/bin, and neither ~/bin nor /usr/bin exist on a clean Ubuntu 16.04 setup, and my $GOPATH isn't inside $PATH (is that really what ppl do? seems kinda sketch...?) so it gave up.

Enumerating every possible bin location is silly af, seems to me the fallbackiest fallback should just ask the user where the binary should be put?

I'll submit a PR for adding ~/.local/bin momentarily.

**Comment by ThomasFreedman** [comment]
Hmmm... Irrespective of "standard" bin folders, either the code or docs should at least provide info on where it *expects*  to install the bin. Even with --verbose no such info is provided. 

I do set GOPATH=/home/$USER/go, and append /usr/local/go/bin to PATH in a 2 line bash script that I source in each user's **.**profile.  I also source that script in my installation script prior to running ipfs-update. 
It will be a simple matter to also append the $GOPATH/bin to that small script sourced in each user's **.**profile, so I will do that.

Thank you for the info, hopefully that will resolve this problem. Since the `install failed, reverting changes...` doesn't remove the copy of the binary used for verification, I just manually copied it to /usr/local/bin/ipfs-update as a workaround.

**Comment by ThomasFreedman** [comment]
Adding $GOPATH/bin to the PATH doesn't resolve the issue. I thought I only added $GOPATH to the PATH but no, I added both the go language bin AND $GOPATH/bin to it and the problem still occurs. $GOPATH is set to /home/$USER/go. Since I'm running my script as root with sudo, permissions aren't the issue.

I verified this very carefully. 

I downloaded version 1.5.2 of ipfs-update, although the "version" option reports 1.5.1 (binary version wasn't updated to match the label on download page)

**Comment by hsanjuan** [comment]
closing as old, and the actual issue was solved. Feel free to open a new issue about the documentation or, better, send a PR to improve it.

**Comment by ThomasFreedman** [comment]
Could you update this issue with a reference to the resolution? Fixed in release # or something to help people that run into this in the future (and so I can update my docs & scripts to reflect the fix)?

**Comment by Stebalien** [comment]
We rewrote the install logic in https://github.com/ipfs/ipfs-update/pull/112. This logic has now been documented in the [README](https://github.com/ipfs/ipfs-update#install-location) and was included in releases v1.5.3 and v1.6.0.

**Comment by LeDechaine** [comment]
This bug is back. Fix:

# /tmp/ipfs-update441531582/ipfs-new version
ipfs version 0.12.0
# ipfs --version
ipfs version 0.11.0
# whereis ipfs
ipfs: /usr/local/bin/ipfs
**# mv /tmp/ipfs-update441531582/ipfs-new /usr/local/bin/ipfs**
# wget https://dist.ipfs.io/fs-repo-migrations/v2.0.2/fs-repo-migrations_v2.0.2_linux-amd64.tar.gz
# tar -xvf fs-repo-migrations_v2.0.2_linux-amd64.tar.gz 
fs-repo-migrations/fs-repo-migrations
**# fs-repo-migrations**
(...)
2022/03/14 05:24:49 Success: fs-repo migrated to version 12.
**# systemctl restart ipfs**

Done.

**Comment by LeDechaine** [comment]
**Augh**, apologies. I thought I had installed it recently, but didn't. It needed to be updated and (ipfs-update) doesn't update itself (yet). I upgraded to the latest version and it (appears) to work ok now.

Still, I hope my comment will help others who haven't updated (...ipfs-update) yet and got this cryptic "QmTFJQ68kaArzsqz2Yjg1yMyEA5TXTfNw6d9wSFhxtBxz2" error, because a few weeks ago even google didn't knew what this error meant.

IPFS is great, by the way. :smile:
